src/Command/Backup/BackupDatabaseCommand.php

```diff
@@ -0,0 +1,161 @@
+
+        $namedPipe = $this->createNamedPipe();
+        $compressProcess = null;
+
+        try {
+            $compressProcess = $this->createCompressProcess($namedPipe, $backupPath);
+            $compressProcess->start();
+
+            $dumpProcess = $this->handler->createDumpProcess($namedPipe);
+            $dumpProcess->run();
+
+            if (false === $dumpProcess->isSuccessful()) {
+                throw new RuntimeException($dumpProcess->getErrorOutput());
+            }
+
+            $compressProcess->wait();
+        } catch (Throwable $throwable) {
+            if (null !== $compressProcess && $compressProcess->isRunning()) {
+                $compressProcess->stop(0);
+            }
+
+            throw $throwable;
+        } finally {
+            // Remove potentially created named pipe.
+            $this->filesystem->remove($namedPipe);
+        }
+
+        $errorOutput = $dumpProcess->getErrorOutput();
+        $io->getErrorStyle()->write($errorOutput);
+
+        $isFailure =
+            false === $dumpProcess->isSuccessful() ||
+            false === $compressProcess->isSuccessful() ||
+            // Some errors like missing PROCESS privilege result in success exit code but with corresponding error.
+            0 === strpos($errorOutput, 'mysqldump: Error: ');
+
+        if ($isFailure) {
+            // Remove the empty file created by the command's output pipe.
+            $this->filesystem->remove($backupPath);
+
+            $io->getErrorStyle()->newLine();
+            $io->error('Failed to backup database.');
+
+            return 1;
+        }
+
+        $io->success('Finished database backup.');
+
+        return 0;
+    }
+
+    private function createNamedPipe(): string
+    {
+        $namedPipe = sprintf('%1$s/%2$s', $this->targetDirectory, bin2hex(random_bytes(32)));
+
+        $process = new Process(['mkfifo', $namedPipe]);
+
+        $process->setPty(false);
+        $process->setTimeout(null);
+        $process->setTty(false);
+
+        $process->run();
+
+        if (false === $process->isSuccessful()) {
+            // Remove potentially created named pipe.
+            $this->filesystem->remove($namedPipe);
+
+            throw new UnexpectedValueException('Failed creating named pipe.');
+        }
+
+        return $namedPipe;
+    }
+
+    private function createCompressProcess(string $namedPipe, string $backupPath): Process
+    {
+        $process = Process::fromShellCommandline(implode(' ', $this->getCompressCommandLineParts()));
+
+        $process->setPty(false);
+        $process->setTimeout(null);
+        $process->setTty(false);
+
+        $process->setEnv([
+            'DB_CONTENT_PIPE' => $namedPipe,
+            'DB_BACKUP_PATH' => $backupPath,
+        ]);
+
+        return $process;
+    }
+
+    private function getCompressCommandLineParts(): array
+    {
+        return ['bzip2', '<', '"$DB_CONTENT_PIPE"', '>', '"$DB_BACKUP_PATH"'];
```

Comment:
Too many lines to directly include above?

Reply:
see added comment.